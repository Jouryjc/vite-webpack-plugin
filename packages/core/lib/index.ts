import parse5 from 'parse5'
import type { Element, Attribute } from 'parse5'
import MagicString from 'magic-string'
import path from 'path'
import { readFile } from 'fs/promises'

const LANG_TAG_NAME = 'lang'
const LANG_TEXT_REG = /^<lang>(.*?)<\/lang>$/

interface LANG_ITEM {
  content: string
  startOffset: number
  endOffset: number
}

interface replaceInfo {
  key: string,
  expression: string
}

/** 外部传入的配置 */
interface transformOptions {
  genKey: (node: Element | Attribute) => string
  genExpression (str: string): string
}

type internalConfig = transformOptions & {
  i18nTransform (node: Element | Attribute): replaceInfo
}

export function parse(source: string) {
  return parse5.parse(source, {
    sourceCodeLocationInfo: true
  })
}

export function createKeyTransformer (config: transformOptions) {
  const { genKey, genExpression } = config
  
  return (node: Element | Attribute): replaceInfo => {
    let res = new Proxy({
      key: '',
      expression: ''
    }, {
      get (target, key) {
        return target[key]
      },
      set (target, key, value) {
        // 不能单独修改表达式
        if (key === 'expression') {
          throw Error('Modification of the expression property alone is not supported')
        }
  
        // 词条key不能为空
        if (value === '') {
          throw Error('The entry key cannot be empty')
        }
  
        target[key] = value
  
        // 产品线这里很多情况，有 {[]} 的、有 {}(apply 传参的)、有直接将 $i 传入到 tpl 中
        target.expression = genExpression(value)
  
        return true
      }
    })
  
    // todo key 的生成规则，这里需要两台规则，分别对应两种接口的数据
    res.key = genKey(node)
  
    return res
  }
  
}

function resolveOptions(options: transformOptions): internalConfig {
  let config = {
    ...options,
    i18nTransform: createKeyTransformer(options)
  }

  return config
}

export function traverse(tree, func: (param: Element) => void) {
  tree = tree ?? []
  let nodes = Array.isArray(tree) ? tree : [tree]

  for (let node of nodes) {
    if (node?.sourceCodeLocation) {
      func(node)
    }

    traverse(node.childNodes, func)
  }
}


/**
 * 判断是否 lang 的父级
 *
 * @param {Element} node
 * @return {*}  {boolean}
 */
function hasLangParents(node: Element): boolean {
  let parent = node.parentNode

  while (parent) {
    if (parent.nodeName === LANG_TAG_NAME) {
      return true
    }

    parent = (parent as Element)?.parentNode
  }

  return false
}

export function createTransformer(
  options: transformOptions
) {
  const { i18nTransform } = resolveOptions(options)

  return async (source: string) => {
    const langMap = new Map<string, string>()

    if (!source) {
      return {
        code: '',
        langMap
      }
    }

    const ast = parse(source)
    const s = new MagicString(source)

    // 递归遍历 ast，找到有 lang 的标签或者属性就做字符串操作
    traverse(ast, (node: Element) => {
      const { sourceCodeLocation } = node

      // tagName === 'lang' 并且没有父级的 lang
      if (node.nodeName === LANG_TAG_NAME && !hasLangParents(node)) {
        const { startOffset, endOffset } = sourceCodeLocation!
        let text = s.slice(startOffset, endOffset)
        const matchers = text.match(LANG_TEXT_REG)
        text = matchers ? matchers[1] : parse5.serialize(node)

        // 替换 tpl 时，替换代码跟key是两个东西，key 代表语言包中的词条 key，expression 表示的是要真实替换到tpl代码中的字符串
        const { key, expression } = i18nTransform(node)
        langMap.set(key, text)

        s.overwrite(startOffset, endOffset, expression)
      }
 
      // attrs 数组里面的项包含 lang
      (node?.attrs || []).forEach(attr => {
        const { value } = attr
        const matchers = value.match(LANG_TEXT_REG)

        if (matchers) {
          // 替换 tpl 时，替换代码跟key是两个东西，key 代表语言包中的词条 key，expression 表示的是要真实替换到tpl代码中的字符串
          const { key, expression } = i18nTransform(attr)
          langMap.set(key, matchers[1])

          s.replace(value, expression)
         
        }
      })
    })

    return {
      code: s.toString(),
      langMap
    };
  }
}

export async function transformFromFile(file: string, options: transformOptions) {
  const source = await readFile(path.resolve(process.cwd(), file), 'utf-8')
  const result = await createTransformer(options)(source)

  return result
}
