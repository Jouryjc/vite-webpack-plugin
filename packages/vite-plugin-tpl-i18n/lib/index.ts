import type { Plugin, ViteDevServer, ConfigEnv } from 'vite'
import type { transformOptions } from 'tpl-i18n-core'
import { createTransformer } from 'tpl-i18n-core'
import toSource from 'tosource'

const tplRegExp = new RegExp('\\.tpl$')

interface PluginOptions extends transformOptions {
  /** 词条写入路径 */
  yamlPath: string

  command: ConfigEnv['command']
}

function tplI18nPlugin (options: PluginOptions): Plugin {
  let server: ViteDevServer
  let { genExpression, genKey } = options
  let transformTpl = createTransformer({
    genKey,
    genExpression
  })

  return {
    name: 'vite:plugin-i18n-tpl',

    configureServer (_server) {
      server = _server
    },

    async transform (source, id) {
      if (!tplRegExp.test(id)) {
        return;
      }
      
      const { moduleGraph } = server
      const moduleInfo = moduleGraph.getModulesByFile(id)!

      // console.log(moduleInfo)

      // tpl 文件的处理通过 core 模块处理
      const { code } = await transformTpl(source)
      console.log(code)
      // 这一层的职责主要是获取关联模块，将 tpl 的转化结果写入到引用者对应的函数块中
      for (let { importers } of moduleInfo) {

      }

      // todo 将词条写入 yaml 文件

      return `export default ${toSource(code)}`
    }
  }
}

export default tplI18nPlugin