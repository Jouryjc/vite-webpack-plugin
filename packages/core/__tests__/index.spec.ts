import { describe, test, expect, beforeAll, beforeEach } from 'vitest'
import { createTransformer, transformFromFile } from '../lib/index'
import path from 'path'

let transform

async function assetTransform (source: string, result: string) {
  const { code } = await transform(source)
  expect(code).toBe(result)
}

describe('tpl i18n no lang', () => {
  beforeAll(() => {
    transform = createTransformer({
      genKey: (node) => `node_test_1`,
      genExpression: (value) => `{[ this.${value} ]}`
    })
  })

  test('empty string', async () => {
    await assetTransform('', '')
  })

  test('blank character tpl', async () => {
    await assetTransform(' ', ' ')
  })

  test('html tpl', async () => {
    await assetTransform(`<html><body><div></div></body></html>`, `<html><body><div></div></body></html>`)
  })

  test('html fragment', async () => {
    await assetTransform('<p>hello</p>', '<p>hello</p>')
  })
})


describe('tpl i18n lang node', () => {
  let index

  beforeEach(() => {
    index = 0
    transform = createTransformer({
      genKey: (node) => {
        index++
        return `node_test_${index}`
      },
      genExpression: (value) => `{[ this.${value} ]}`
    })
  })

  test('html tpl with lang', async () => {
    const tpl = `<html><body><lang>中文</lang></body></html>`
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('<html><body>{[ this.node_test_1 ]}</body></html>')
    const map = {
      'node_test_1': '中文'
    }
    expect(langMap).toEqual(map)
  })

  test('html fragment with lang', async () => {
    const tpl = '<lang>汉字</lang><div>content<lang>内部用法</lang></div>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('{[ this.node_test_1 ]}<div>content{[ this.node_test_2 ]}</div>')
    const map = {
      'node_test_1': '汉字',
      'node_test_2': '内部用法'
    }
    expect(langMap).toEqual(map)
  })

  test('sibling lang', async () => {
    const tpl = '<lang>汉字</lang><lang>内部用法</lang>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('{[ this.node_test_1 ]}{[ this.node_test_2 ]}')
    const map = {
      'node_test_1': '汉字',
      'node_test_2': '内部用法'
    }
    expect(langMap).toEqual(map)
  })
  
  test('illegal lang', async () => {
    const tpl = '<div><lang>汉字</div></lang>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('<div>{[ this.node_test_1 ]}</div></lang>')
    const map = {
      'node_test_1': '汉字'
    }
    expect(langMap).toEqual(map)
  })

  test('test nested in lang tag', async () => {
    const tpl = '<lang>嵌套<input type="button" />尾部</lang>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('{[ this.node_test_1 ]}')
    const map = {
      'node_test_1': '嵌套<input type="button" />尾部'
    }
    expect(langMap).toEqual(map)
  })

  test('nested lang tag', async () => {
    const tpl = '<lang>头部<lang>中间</lang>尾部</lang>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('{[ this.node_test_1 ]}')
    const map = {
      'node_test_1': '头部<lang>中间</lang>尾部'
    }
    expect(langMap).toEqual(map)
  })

  test('empty lang tag', async () => {
    const tpl = '<lang></lang>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('{[ this.node_test_1 ]}')
    const map = {
      'node_test_1': ''
    }
    expect(langMap).toEqual(map)
  })

  test('special char', async () => {
    const tpl = '<lang>确定{#noun#}</lang>'
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('{[ this.node_test_1 ]}')
    const map = {
      'node_test_1': '确定{#noun#}'
    }
    expect(langMap).toEqual(map)
  })
})

describe('i18n lang attributes', () => {
  let index

  beforeEach(() => {
    index = 0
    transform = createTransformer({
      genKey: (node) => {
        index++
        return `node_test_${index}`
      },
      genExpression: (value) => `{[ this.${value} ]}`
    })
  })

  test('single attr', async () => {
    const tpl = `<input placeholder="<lang>请输入</lang>" />`
    const { code, langMap } = await transform(tpl)
    expect(code).toMatch('<input placeholder="{[ this.node_test_1 ]}" />')
    const map = {
      'node_test_1': '请输入'
    }
    expect(langMap).toEqual(map)
  })

  test('empty lang attr', async () => {
    const tpl = `<input placeholder="<lang></lang>" />`
    const { code, langMap } = await transform(tpl)
    expect(code).toMatch('<input placeholder="{[ this.node_test_1 ]}" />')
    const map = {
      'node_test_1': ''
    }
    expect(langMap).toEqual(map)
  })

  test('multi lang attr', async () => {
    const tpl = `<input placeholder="<lang>请输入</lang>" data-tip="<lang>输入整数</lang>" />`
    const { code, langMap } = await transform(tpl)
    expect(code).toBe('<input placeholder="{[ this.node_test_1 ]}" data-tip="{[ this.node_test_2 ]}" />')
    const map = {
      'node_test_1': '请输入',
      'node_test_2': '输入整数'
    }
    expect(langMap).toEqual(map)
  })

  test('single quota', async () => {
    const tpl = `<input placeholder='<lang>请输入</lang>' />`
    const { code, langMap } = await transform(tpl)
    expect(code).toBe(`<input placeholder='{[ this.node_test_1 ]}' />`)
    const map = {
      'node_test_1': '请输入'
    }
    expect(langMap).toEqual(map)
  })
})

describe ('transform file', () => {

  test('transform from file', async () => {
    const demoFile = path.resolve(__dirname, './demo/test.tpl')
    let index = 0
    const result = await transformFromFile(demoFile, {
      genKey: (node) => {
        index++
        return `node_test_${index}`
      },
      genExpression: (value) => `{[ this.${value} ]}`
    })

    expect(result.code).toMatchSnapshot()
    expect(result.langMap).toEqual({
        'node_test_1': 'hello',
        'node_test_2': '请输入'
    })
  })
})