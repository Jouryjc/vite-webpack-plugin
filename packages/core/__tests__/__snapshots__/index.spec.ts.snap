// Vitest Snapshot v1

exports[`transform file > transform from file 1`] = `
"<div>
  {[ this.node_test_1 ]}
  <input placeholder=\\"{[ this.node_test_2 ]}\\" />
</div>"
`;
